### PR TITLE
[BugFix] fix lake pk compaction invalid memory access

### PR DESCRIPTION
**Pull request by luohaha**
Why I'm doing:
When do compaction and load state, `get_each_segment_iterator` can't use a stat in a stack. otherwise it will access invalid memory. It will cause crash like:

```
*** Aborted at 1704856337 (unix time) try "date -d @1704856337" if you are using GNU date ***
PC: @          0x51e021b starrocks::lake::TabletReader::do_get_next()
*** SIGSEGV (@0x0) received by PID 15529 (TID 0x7f8315335700) from PID 0; stack trace: ***
    @          0x65f27c2 google::(anonymous namespace)::FailureSignalHandler()
    @     0x7f83ce32cb20 (unknown)
    @          0x51e021b starrocks::lake::TabletReader::do_get_next()
    @          0x5214800 starrocks::lake::VerticalCompactionTask::compact_column_group()
    @          0x5215705 starrocks::lake::VerticalCompactionTask::execute()
    @          0x51b33c1 starrocks::lake::CompactionScheduler::do_compaction()
    @          0x51b3de5 starrocks::lake::CompactionScheduler::thread_task()
    @          0x2dde50a starrocks::ThreadPool::dispatch_thread()
    @          0x2dd8f7a starrocks::Thread::supervise_thread()
    @     0x7f83ce32214a start_thread
    @     0x7f83cd6acdc3 __GI___clone
    @                0x0 (unknown)
```

What I'm doing:
`get_each_segment_iterator` with gloabl stat.

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wyb** [comment]
use once flag like local update compaction state?

**Comment by wanpengfei-git** [comment]
PR was manually closed.

**Comment by luohaha** [comment]
> use once flag like local update compaction state?

But we need to load compaction state one segment by one segment for saving memory.

**Comment by wanpengfei-git** [comment]
PR was manually closed.

**Review by TszKitLo40 — APPROVED** [review]

**Review by tracymacding — APPROVED** [review]

**Review by xiangguangyxg — APPROVED** [review]

**Review by wyb — DISMISSED** [review]

File: be/src/storage/lake/update_compaction_state.cpp [hunk 1]
```diff
@@ -29,6 +29,8 @@ CompactionState::~CompactionState() {
     if (_update_manager != nullptr) {
         _update_manager->compaction_state_mem_tracker()->release(_memory_usage);
     }
+    // make sure segment iters release before `_stats`
```
Comment (sduzh) [inline be/src/storage/lake/update_compaction_state.cpp:32]: If you want the `_segment_iters` to be destructed before `_stats`, why not just declare `_stats` before the `_segment_iters`?

Comment (luohaha) [reply be/src/storage/lake/update_compaction_state.cpp]: make sense, done.

**Review by wyb — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**